### PR TITLE
New nrow_subset_linter

### PR DESCRIPTION
**Pull request by MichaelChirico**
Part of #884

No hits on {lintr}

Will convert some of the TODOs as follow-up issues, I was surprised how many hits this simple version gets already though.

### CODE REVIEW

**Comment by MichaelChirico** [comment]
Will convert the TODO to new issues after initial review to satisfy `commented_code_linter()`.

File: tests/testthat/test-nrow_subset_linter.R [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+#   lintr::expect_lint(
+#     "x[y == z, ]",
+#     "Use arithmetic to count the number of rows satisfying a condition",
+#     nrow_subset_linter
+#   )
+# })
+
+# test_that("nrow_subset_linter blocks dplyr::filter() cases", {
+#   lintr::expect_lint(
+#     "x %>% filter(y == z) %>% nrow()",
```
Comment (AshesITR) [inline tests/testthat/test-nrow_subset_linter.R:46]: This one I have seen many times.

File: tests/testthat/test-nrow_subset_linter.R [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+#   # data.frame subsetting (NB: replacement doesn't use na.rm = TRUE)
+#   lintr::expect_lint(
+#     "nrow(x[x$y == x$z, ])",
+#     "Use arithmetic to count the number of rows satisfying a condition",
+#     nrow_subset_linter
+#   )
+
+#   # data.table subsetting (NB: replacement needs na.rm = TRUE)
+#   lintr::expect_lint(
+#     "x[y == z, ]",
```
Comment (AshesITR) [inline tests/testthat/test-nrow_subset_linter.R:38]: Probably you want `.N` in j= here.

File: tests/testthat/test-nrow_subset_linter.R (hunk 1 above)
Comment (MichaelChirico) [reply tests/testthat/test-nrow_subset_linter.R:46]: #2313 and #2314 are the follow-ups here

File: tests/testthat/test-nrow_subset_linter.R (hunk 2 above)
Comment (MichaelChirico) [reply tests/testthat/test-nrow_subset_linter.R:38]: I think I meant it to be `nrow(x[y == z, ])`... anyway #2315 is the follow-up.

File: R/nrow_subset_linter.R [hunk 3]
```diff
@@ -0,0 +1,43 @@
+#' Block usage of `nrow(subset(x, .))`
+#'
+#' Using `nrow(subset(x, condition))` to count the instances where `condition`
+#'   applies inefficiently requires doing a full subset of `x` just to
+#'   count the number of rows in the resulting subset.
+#' There are a number of equivalent expressions that don't require the full
+#'   subset, e.g. `with(x, sum(condition))` (or, more generically,
+#'   `with(x, sum(condition, na.rm = TRUE))`).
+#' The same can be said of other versions of this like
```
Comment (AshesITR) [inline R/nrow_subset_linter.R:9]: We should be clear that this is currently not detected.

File: R/nrow_subset_linter.R [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+nrow_subset_linter <- make_linter_from_xpath(
+  xpath = "
+  //SYMBOL_FUNCTION_CALL[text() = 'subset']
+    /parent::expr
+    /parent::expr
+    /parent::expr[expr/SYMBOL_FUNCTION_CALL[text() = 'nrow']]
+  ",
+  lint_message = paste(
+    "Use arithmetic to count the number of rows satisfying a condition,",
+    "rather than fully subsetting the table and counting the resulting rows.",
```
Comment (AshesITR) [inline R/nrow_subset_linter.R:38]: table -> data frame

File: R/nrow_subset_linter.R [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    "For example, replace nrow(subset(x, is_treatment))",
+    "with sum(x$is_treatment). NB: use na.rm = TRUE if `is_treatment` has",
```
Comment (AshesITR) [inline R/nrow_subset_linter.R:40]: `sum(x$is_treatment, na.rm = TRUE)` seems more concise and easy enough to understand?

File: inst/lintr/linters.csv [hunk 6]
```diff
@@ -57,6 +57,7 @@ namespace_linter,correctness robustness configurable executing
 nested_ifelse_linter,efficiency readability
 no_tab_linter,style consistency deprecated
 nonportable_path_linter,robustness best_practices configurable
+nrow_subset_linter,efficiency consistency readability best_practices
```
Comment (AshesITR) [inline inst/lintr/linters.csv:60]: I don't really see how this helps readability tbh.

File: R/nrow_subset_linter.R (hunk 3 above)
Comment (MichaelChirico) [reply R/nrow_subset_linter.R]: I think I wrote that with the intention of having it implemented but didn't follow through. Dropping the reference for now.

File: inst/lintr/linters.csv (hunk 6 above)
Comment (MichaelChirico) [reply inst/lintr/linters.csv]: Fair... I think of it in a context where I often see things like `if (nrow(subset(x, many & con & ditions)) > 0) { x <- subset(x, many & con & ditions) }` where there's a readability issue but it's not strictly about the `nrow(subset(.))` call.

I removed the tag.

**Comment by AshesITR** [comment]
Could use a lot more testing, especially negatives, vectorization and metadata.

**Review by AshesITR — APPROVED** [review]

**Review by AshesITR — APPROVED** [review]

**Pull request merged**